cirq/work/collector.py

```diff
@@ -0,0 +1,164 @@
+                                *,
+                                concurrency: int = 2,
+                                max_total_samples: Optional[int] = None
+                               ) -> None:
+    """Concurrently collects samples from a simulator or hardware sampler.
+
+    Examples:
+
+        ```
+        collector = cirq.PauliStringCollector(...)
+        await cirq.async_collect_samples(
+            sampler=cirq.Simulator(),
+            collector=collector,
+            concurrency=3)
+        print(collector.estimated_energy())
+        ```
+
+    See Also:
+
+        Python 3 documentation "Coroutines and Tasks"
+        https://docs.python.org/3/library/asyncio-task.html
+
+    Args:
+        collector: Determines which circuits to sample next, and processes the
+            resulting samples. This object will be mutated.
+        sampler: The simulator or hardware to take samples from.
+        concurrency: Desired number of sampling jobs to have in flight at any
+            given time.
+        max_total_samples: Optional limit on the maximum number of samples to
+            collect.
+
+    Returns:
+        The collector's result after all desired samples have been collected.
+    """
+    pool = work_pool.CompletionOrderedAsyncWorkPool()
+    remaining_samples = (np.infty
+                         if max_total_samples is None else max_total_samples)
+
+    async def _start_async_job(job):
+        return job, await sampler.async_sample(job.circuit,
+                                               repetitions=job.repetitions)
+
+    # Keep dispatching and processing work.
+    while True:
+        # Fill up the work pool.
+        while (remaining_samples > 0 and pool.num_uncollected < concurrency):
+            new_jobs = _flatten_jobs(collector.next_job())
+
+            # If no jobs were given, stop asking until something completes.
+            if not new_jobs:
+                break
+
+            # Start new sampling jobs.
+            for new_job in new_jobs:
+                remaining_samples -= new_job.repetitions
+                pool.include_work(_start_async_job(new_job))
+
+        # If no jobs were started or running, we're in a steady state. Halt.
+        if not pool.num_uncollected:
+            break
+
+        # Forward next job result from pool.
+        done_job, done_val = await pool.__anext__()
+        collector.on_job_result(done_job, done_val)
+
+
+def _flatten_jobs(
+        given: Union[None, CircuitSampleJob, Iterable[CircuitSampleJob]]):
+    out = []  # type: List[CircuitSampleJob]
+    _flatten_jobs_helper(out, given)
+    return out
+
+
+def _flatten_jobs_helper(
+        out: List[CircuitSampleJob],
+        given: Union[None, CircuitSampleJob, Iterable[CircuitSampleJob]]):
+    if isinstance(given, CircuitSampleJob):
+        out.append(given)
+    elif given is not None:
+        for item in given:
+            _flatten_jobs_helper(out, item)
```

Comment:
This implementation allows nested iterables of `CircuitSampleJob`, which the type annotation does not allow. Did you intend to allow nested interables but can't express it in the types?

Reply:
Updated the type annotation.